src/main/src/main/java/org/geoserver/ows/StylePublisher.java

```diff
@@ -0,0 +1,123 @@
+
+    /**
+     * Creates the new file publisher.
+     * 
+     * @param loader The loader used to locate files.
+     */
+    public StylePublisher(Catalog catalog) {
+        this.catalog = catalog;
+    }
+
+    // @Override
+    // protected void initServletContext(ServletContext servletContext) {
+    // this.scloader = new ServletContextResourceLoader(servletContext);
+    // }
+
+    @Override
+    protected URL getUrl(HttpServletRequest request) throws IOException {
+        String ctxPath = request.getContextPath();
+        String reqPath = request.getRequestURI();
+        reqPath = URLDecoder.decode(reqPath, "UTF-8");
+        reqPath = reqPath.substring(ctxPath.length());
+
+        if ((reqPath.length() > 1) && reqPath.startsWith("/")) {
+            reqPath = reqPath.substring(1);
+        }
+
+        // sigh, in order to serve the file we have to open it 2 times
+        // 1) to determine its mime type
+        // 2) to determine its encoding and really serve it
+        // we can't coalish 1) because we don't have a way to give jmimemagic the bytes at the
+        // beginning of the file without disabling extension quick matching
+
+        // load the file
+        Resource resource;
+
+        // if the file is like "styles/*.png" then look in the global styles folder
+        if (reqPath.startsWith("styles/")) {
+            // tokenize string
+            // check # of tokens
+            // if 2 => global style
+            // if 3 => worksResource resource;paced style
+
+            String[] tokens = reqPath.split("/");
+            int count = tokens.length;
+            GeoServerResourceLoader resources = catalog.getResourceLoader();
+            GeoServerDataDirectory data = new GeoServerDataDirectory(resources);
+            if (count == 2) {
+                // global style
+                String stylePath = tokens[1];
+                resource = data.getStyles(stylePath);
+            } else if (count == 3) {
+                // workspaced style
+                // same thing as above but using new path we just constructed
+                String temp = "workspaces/" + tokens[1] + "/styles/" + tokens[2];
+                String wsName = tokens[1];
+                String stylePath = tokens[2];
+
+                WorkspaceInfo workspace = catalog.getWorkspaceByName(wsName);
+                if (workspace == null) {
+                    throw new IllegalArgumentException("Workspace not found");
+                }
+                resource = data.getStyles(workspace, stylePath);
+                if (resource.getType() == Type.UNDEFINED) {
+                    // workspace style not found, try the global styles directory as a fallback
+                    resource = data.getStyles(stylePath);
+                }
+            } else {
+                // something went wrong :(
+                throw new IllegalStateException("Unexpected number of tokens in reqest path.");
```

Comment:
This error message is unlikely to help a normal user (who does not think in terms of tokens unless they are playing in an arcade)